### PR TITLE
fix(components): [input] fix textarea autosize

### LINKED ISSUE #11375: [Style] [input] 在火狐浏览器中，自适应文本域（使用了autosize属性的输入框）组件无法自动调整高度
Bug Type: **`Style`**

## Environment

- Vue Version: `3.2.45`
- Element Plus Version: `2.2.28`
- Browser / OS: `UserAgent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/109.0`
- Build Tool: `Other`

### Related Component

- `el-input`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtaW5wdXRcbiAgICB2LW1vZGVsPVwidGV4dGFyZWExXCJcbiAgICBhdXRvc2l6ZVxuICAgIHR5cGU9XCJ0ZXh0YXJlYVwiXG4gICAgcGxhY2Vob2xkZXI9XCJQbGVhc2UgaW5wdXRcIlxuICAvPlxuICA8ZGl2IHN0eWxlPVwibWFyZ2luOiAyMHB4IDBcIiAvPlxuICA8ZWwtaW5wdXRcbiAgICB2LW1vZGVsPVwidGV4dGFyZWEyXCJcbiAgICA6YXV0b3NpemU9XCJ7IG1pblJvd3M6IDIsIG1heFJvd3M6IDQgfVwiXG4gICAgdHlwZT1cInRleHRhcmVhXCJcbiAgICBwbGFjZWhvbGRlcj1cIlBsZWFzZSBpbnB1dFwiXG4gIC8+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IGxhbmc9XCJ0c1wiIHNldHVwPlxuaW1wb3J0IHsgcmVmIH0gZnJvbSAndnVlJ1xuY29uc3QgdGV4dGFyZWExID0gcmVmKCcnKVxuY29uc3QgdGV4dGFyZWEyID0gcmVmKCcnKVxuPC9zY3JpcHQ+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)

### Steps to reproduce

使用火狐浏览器（Windows版或安卓版）打开上述最小复现环境的链接，在右侧的任意一个输入框中输入任意四行字符。

### What is Expected?

该输入框在输入任意四行字符后，高度自动调整为4行。

### What is actually happening?

该输入框在输入任意四行字符时，高度没有任何变化。

### Additional comments

下面的链接是Element官网针对自适应文本域的示例：
https://element-plus.gitee.io/zh-CN/component/input.html#%E8%87%AA%E9%80%82%E5%BA%94%E6%96%87%E6%9C%AC%E5%9F%9F

在该示例中，任意一个输入框在输入任意四行字符后，高度确实能够自动调整为4行。这是符合期望的。

但是在最小复现环境链接对应的页面中就不是这样了。点击上述示例的“在 Playground 中编辑”按钮，跳转的链接便是上面的最小复现环境链接。在该页面中，任意一个输入框在输入任意四行字符后，高度不会自动调整。这显然不符合期望。

在Edge和Chrome中，并没有上述问题；但在Windows版Firefox和安卓版Firefox中，均有上述问题。

简而言之，在火狐浏览器中，el-input组件的autosize属性的minRows能够生效，但是当输入的文本行数超过minRows时，该组件的高度无法自动调整。并且，该组件在官网的“自适应文本域”示例的表现符合期望，但在“在 Playground 中编辑”页面中的表现不符合期望。

### PR DESCRIPTION
**Pull request by acyza**
Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

fix #11375 
这应该是firefox的bug，firefox给`<textarea />`给与样式`overflow:hidden;position:absolute;right:0;`并设置value值后立即获取scrollHeight得到的值是错误的，再获取一次可以得到正确的值。

### CODE REVIEW

File: packages/components/input/src/utils.ts [hunk 1]
```diff
@@ -78,6 +78,9 @@ export function calcTextareaHeight(
   hiddenTextarea.setAttribute('style', `${contextStyle};${HIDDEN_STYLE}`)
   hiddenTextarea.value = targetElement.value || targetElement.placeholder || ''
 
+  //修复firefox高度获取错误
+  hiddenTextarea.scrollHeight
```
Comment (chenxch) [inline packages/components/input/src/utils.ts:82]: Perhaps there is a more secure solution?

Comment (acyza) [reply packages/components/input/src/utils.ts:82]: [Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICB2YWx1ZeWAvOKAnHt7dmFsdWV9feKAne+8jOiOt+WPluWIsOeahHNjcm9sbEhlaWdodHt7aGVpZ2h0fX1cbiAgPGJyPlxuICB2YWx1ZeWAvOKAnHt7dmFsdWUxfX3igJ3vvIzojrflj5bliLDnmoRzY3JvbGxIZWlnaHR7e2hlaWdodDF9fVxuICA8YnI+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IGxhbmc9XCJ0c1wiIHNldHVwPlxuaW1wb3J0IHsgcmVmLG9uTW91bnRlZCB9IGZyb20gJ3Z1ZSdcbmNvbnN0IGhlaWdodCA9IHJlZigpXG5jb25zdCB2YWx1ZSA9IHJlZigpXG5jb25zdCBoZWlnaHQxID0gcmVmKClcbmNvbnN0IHZhbHVlMSA9IHJlZigpXG5vbk1vdW50ZWQoKCk9Pntcblx0Y29uc3QgdGV4dGFyZWEgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCd0ZXh0YXJlYScpXG4gIC8vSElEREVOX1NUWUxF5Lit55qE6YOo5YiG5qC35byPXG4gIHRleHRhcmVhLnNldEF0dHJpYnV0ZSgnc3R5bGUnLCdvdmVyZmxvdzpoaWRkZW4gIWltcG9ydGFudDtwb3NpdGlvbjphYnNvbHV0ZSAhaW1wb3J0YW50O3JpZ2h0OjAgIWltcG9ydGFudDsnKVxuICBkb2N1bWVudC5ib2R5LmFwcGVuZENoaWxkKHRleHRhcmVhKVxuICB2YWx1ZS52YWx1ZSA9IHRleHRhcmVhLnZhbHVlPVxuICAgIGBhXG4gICAgYlxuICAgIGNcbiAgICBkXG4gICAgZVxuICAgIGZcbiAgICBnYFxuICAvL3RleHRhcmVhLnNjcm9sbEhlaWdodFxuICBoZWlnaHQudmFsdWUgPSB0ZXh0YXJlYS5zY3JvbGxIZWlnaHRcbiAgdmFsdWUxLnZhbHVlID0gdGV4dGFyZWEudmFsdWU9JydcbiAgaGVpZ2h0MS52YWx1ZT10ZXh0YXJlYS5zY3JvbGxIZWlnaHRcbn0pXG48L3NjcmlwdD5cbiIsImltcG9ydF9tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHt9XG59IiwiX28iOnt9fQ==)
@chenxch 或者改这几个样式？
https://github.com/element-plus/element-plus/blob/dev/packages/components/input/src/utils.ts#L5

Comment (acyza) [reply packages/components/input/src/utils.ts:82]: 这种浏览器自身的bug好像很难给它做的很完美。

**Comment by acyza** [comment]
[复现scrollHeight连续获取两次得到的值不同](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8aDE+5L2/55So54Gr54uQ5omT5byAPC9oMT5cbiAg6L+e57ut5Lik5qyh6I635Y+Wc2Nyb2xsSGVpZ2h0PGJyPlxuICDnrKzkuIDmrKHojrflj5bliLDnmoRzY3JvbGxIZWlnaHTvvJp7e2hlaWdodDF9fTxicj5cbiAg56ys5LqM5qyh6I635Y+W5Yiw55qEc2Nyb2xsSGVpZ2h077yae3toZWlnaHQyfX08YnI+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IGxhbmc9XCJ0c1wiIHNldHVwPlxuaW1wb3J0IHsgcmVmLG9uTW91bnRlZCB9IGZyb20gJ3Z1ZSdcbmNvbnN0IGhlaWdodDEgPSByZWYoKVxuY29uc3QgaGVpZ2h0MiA9IHJlZigpXG5vbk1vdW50ZWQoKCk9Pntcblx0Y29uc3QgdGV4dGFyZWEgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCd0ZXh0YXJlYScpXG4gIC8qXG4gICAqIGxpbmsgaHR0cHM6Ly9naXRodWIuY29tL2VsZW1lbnQtcGx1cy9lbGVtZW50LXBsdXMvYmxvYi9kZXYvcGFja2FnZXMvY29tcG9uZW50cy9pbnB1dC9zcmMvdXRpbHMudHMjTDVcbiAgICogSElEREVOX1NUWUxF5Lit55qE6YOo5YiG5qC35byPXG4gICAqL1xuICB0ZXh0YXJlYS5zZXRBdHRyaWJ1dGUoJ3N0eWxlJywnb3ZlcmZsb3c6aGlkZGVuO3Bvc2l0aW9uOmFic29sdXRlO3JpZ2h0OjA7JylcbiAgZG9jdW1lbnQuYm9keS5hcHBlbmRDaGlsZCh0ZXh0YXJlYSlcbiAgLy/nu5l0ZXh0YXJlYeS4gOS4qjfooYznmoTmlofmnKxcbiAgdGV4dGFyZWEudmFsdWU9XG5cdCBgYVxuICAgIGJcbiAgICBjXG4gICAgZFxuICAgIGVcbiAgICBmXG4gICAgZ2BcbiAgLy/ov57nu63ojrflj5ZzY3JvbGxIZWlnaHRcbiAgY29uc3QgdmFsdWUxID0gdGV4dGFyZWEuc2Nyb2xsSGVpZ2h0XG4gIGNvbnN0IHZhbHVlMiA9IHRleHRhcmVhLnNjcm9sbEhlaWdodFxuICBoZWlnaHQxLnZhbHVlID0gdmFsdWUxXG4gIGhlaWdodDIudmFsdWUgPSB2YWx1ZTJcbn0pXG48L3NjcmlwdD5cbiIsImltcG9ydF9tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHt9XG59IiwiX28iOnt9fQ==)

>firefox 版本109.0

**Comment by acyza** [comment]
新方案说明：
>overflow:hidden应该是为了避免滚动条的宽度对文本行数的影响，而firefox的默认滚动条是不占用宽度的，所以用::-webkit-scrollbar伪类去掉除firefox外的滚动条，firefox的滚动条可以用`scrollbar-width:none`隐藏，但通过此方法隐藏掉firefox的滚动条后bug依旧。

**Comment by chenxch** [comment]
@acyza 做了几个简单场景的测试 chrome和firefox暂时没发现问题，但是取消overflow改为修改webkit-scrollbar 要评估一下是否会有其他问题。
@element-plus/backers

**Comment by chenxch** [comment]
https://github.com/element-plus/element-plus/blob/ad7ba95a2cca75cd036c6b3f82a67b0bb6c08a02/packages/components/input/src/utils.ts#L109-L110

我尝试移除这两行代码也是可以的

**Comment by acyza** [comment]
> https://github.com/element-plus/element-plus/blob/ad7ba95a2cca75cd036c6b3f82a67b0bb6c08a02/packages/components/input/src/utils.ts#L109-L110
> 
> 我尝试移除这两行代码也是可以的


```<template>
  <el-input
    v-model="textarea1"
    autosize
    type="textarea"
    placeholder="Please input"
  />
</template>

<script lang="ts" setup>
import { ref } from 'vue'
const textarea1 = ref(’‘)
textarea1.value= `1
  2
  3
  4
  5
  6
  7
`
</script>
```
删除这两行像这样的会有点问题

https://user-images.githubusercontent.com/101238421/216681536-e7e2b7a5-709d-4a9b-aa43-80259262bf72.mp4
一开始只有一行的高度，改变文本内容突然变大。

**Comment by chenxch** [comment]
这个表现不是和目前的一致么，你是认为这个只能是一行么，你的是字符串类型，带有换行符哦

**Comment by acyza** [comment]
是的，用 [link](#issuecomment-1416203224)这个方案当v-model传入一个带换行符的字符串仍然有可能触发firefox的这个bug。

> 这个表现不是和目前的一致么，你是认为这个只能是一行么，你的是字符串模板，带有换行符哦

**Comment by acyza** [comment]
此bug在浏览器窗口高度较小，例如小于body内容高度时不会触发。

https://user-images.githubusercontent.com/101238421/216874360-b1ae40aa-0f3d-4ba5-a5a0-b0fa70f1baa8.mp4

**Review by jw-foss — COMMENTED** [review]
@acyza I recall that we have a utility method in our codebase called `isFirefox`, can we use that to conditionally add this class to the element?

**Comment by acyza** [comment]
@jw-foss 有isFirefox的话就不用这么麻烦了， 因为firefox的滚动条是不占宽度的，直接针对Firefox去掉overflow:hidden就好。

**Comment by acyza** [comment]
https://github.com/element-plus/element-plus/pull/11376#pullrequestreview-1294706327
@jw-foss 那个hidden-webkit-scrollbar是代替`overflow:hidden`给firefox以外的清除滚动条用的，不是给firefox用的。
https://developer.mozilla.org/zh-CN/docs/Web/CSS/::-webkit-scrollbar#%E6%B5%8F%E8%A7%88%E5%99%A8%E5%85%BC%E5%AE%B9%E6%80%A7

**Review by jw-foss — APPROVED** [review]
In code way, it LGTM, /cc @chenxch could you help verifying if the changes are valid?

**Comment by chenxch** [comment]
> In code way, it LGTM, /cc @chenxch could you help verifying if the changes are valid?

OK, leave it to me for a second confirmation.

**Review by chenxch — APPROVED** [review]

**Comment by chenxch** [comment]
@acyza Thank you for your contribution.👍

**Pull request merged**